### ISSUE TITLE
destroy > systematic DependencyViolation

### ISSUE THREAD

**Issue by frntn** [hashicorp/terraform#497]
When trying to `terraform destroy` while running instance(s) **with public ip(s)** on a subnet inside a VPC with an Internet Gateway associated with it, each time I get the following error when trying to destroy the igw : 
```
[...]

aws_internet_gateway.gw-to-internet: Destroying...
aws_internet_gateway.gw-to-internet: Error: Network vpc-xxxxxxxx has some mapped public address(es). Please unmap those public address(es) before detaching the gateway. (DependencyViolation)

[...]

Error applying plan:

1 error(s) occurred:

* Network vpc-xxxxxxx has some mapped public address(es). Please unmap those public address(es) before detaching the gateway. (DependencyViolation)
```

A full `repro.tf` is available here : https://gist.github.com/frntn/12126851cdc2ef1cdbaf

If this is impossible, then the internet gateway destruction attempt should occur **after** every public addresses are detached.

May be wait for *at least* : 
- each known ec2 instances with `associate_public_ip_address` (have to add it as attribute of `aws_instance`) 
- and each ec2 instances inside a subnet with `map_public_ip_on_launch` set to true (have to add it as attribute of `aws_subnet`).

But you may have better way to implement this dependency. Just saying.. :)

--
version: terraform v0.3.1
platform: linux 64bits (debian wheezy)

**Comment by blakeblackshear** [comment]
This is still an issue for me in 0.3.6

**Comment by sbward** [comment]
+1 I'm also seeing this exact issue

**Comment by chipi** [comment]
+1 ... really anoying

**Comment by korya** [comment]
Faced the same issue.

My workaround is to define an explicit dependency of the instances on VPC's gateway.

In my setup there are 4 instances in a VPC (`www_a`, `www_c`, `db_a`, `db_c`):

```tf
resource "aws_instance" "www_a" {
  ...
  depends_on = ["aws_internet_gateway.prd_plntr"]
}
resource "aws_instance" "www_c" {
  ...
  depends_on = ["aws_internet_gateway.prd_plntr"]
}
resource "aws_instance" "db_a" {
  ...
  depends_on = ["aws_internet_gateway.prd_plntr"]
}
resource "aws_instance" "db_a" {
  ...
  depends_on = ["aws_internet_gateway.prd_plntr"]
}

resource "aws_internet_gateway" "my_gw" {
  vpc_id "${aws_vpc.prd_plntr}"
}
```

![No explicit dependency](https://cloud.githubusercontent.com/assets/148461/5921728/f7411256-a613-11e4-874e-011eef61dbb2.png)

![Explicit dependency](https://cloud.githubusercontent.com/assets/148461/5921732/01551012-a614-11e4-8658-66a9008d00c0.png)

I believe this dependency should be set implicitly by terraform.

**Comment by radeksimko** [comment]
Here's a smallest example for reproducing the problem:
```ruby
resource "aws_vpc" "default" {
  cidr_block = "10.10.0.0/16"
}

resource "aws_internet_gateway" "default" {
  vpc_id = "${aws_vpc.default.id}"
}

resource "aws_subnet" "public" {
  vpc_id = "${aws_vpc.default.id}"
  cidr_block = "10.10.0.0/24"
  availability_zone = "eu-west-1a"
}

resource "aws_instance" "nat" {
  ami = "ami-14913f63"
  availability_zone = "eu-west-1a"
  instance_type = "t2.micro"
  key_name = "radeksimko"
  subnet_id = "${aws_subnet.public.id}"
  associate_public_ip_address = true
}
```

If I set `associate_public_ip_address = false` or remove the internet gateway, the issue disappears.

It is reproducible with the very latest version `v0.3.7.dev (26156981d7b67dce7a033ffea94aea5370c09c58)` by just running

```
terraform apply
terraform destroy
```

Simplest workaround is just to run `destroy` twice.

**Comment by clstokes** [comment]
Running into this too.

:+1:

**Comment by catsby** [comment]
This resurfaced recently in #1203 . The workaround that resulted for this particular issue was to add a `depends_on` clause to the instance (as @korya suggests above). To use @radeksimko 's config example:

```
resource "aws_vpc" "default" {
  cidr_block = "10.10.0.0/16"
}

resource "aws_internet_gateway" "default" {
  vpc_id = "${aws_vpc.default.id}"
}

resource "aws_subnet" "public" {
  vpc_id = "${aws_vpc.default.id}"
  cidr_block = "10.10.0.0/24"
  availability_zone = "eu-west-1a"
}

resource "aws_instance" "nat" {
  ami = "ami-14913f63"
  availability_zone = "eu-west-1a"
  instance_type = "t2.micro"
  key_name = "radeksimko"
  subnet_id = "${aws_subnet.public.id}"
  associate_public_ip_address = true
  depends_on = ["aws_internet_gateway.default"] # explicit dependency 
}
```

Terraform is trying to move on and delete other things, but AWS API is still working on destroying things so as to not violate dependencies. Documentation was added in #1189 to document this, but that change hasn't been deployed to terraform.io yet. 

I'm not sure what else can be done here. cc @phinze for ideas about implicitly setting that dependency(?)

**Comment by zxjinn** [comment]
The way I worked around this specific issue was: I associated EIP's with my instances instead of giving them public IPs. When Terraform goes to destroy the VPC it disassociates the EIPs from the instances cleanly, then destroys the EIP. I have a small number of publicly facing instances so this works for me.

Additionally, there is no [extra charge](http://aws.amazon.com/ec2/pricing/#Elastic_IP_Addresses) associated with EIPs as long as you are using them, however you are [limited to five EIPs per region per account](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Appendix_Limits.html).
